### PR TITLE
vmalert-974: fix order for labels templating

### PR DESCRIPTION
**Pull request by hagen1778**
The change fixes bug caused by https://github.com/VictoriaMetrics/VictoriaMetrics/commit/3adf8c5a6f4a8eea5f2025dd53ba975d42f8b8d1.

https://github.com/VictoriaMetrics/VictoriaMetrics/issues/974

------------------

@f41gh7 pls see the test cases in PR. I'd appreciate your review and feel free to suggest additional test cases we may want to check.

### CODE REVIEW

File: app/vmalert/alerting.go [hunk 1]
```diff
@@ -141,11 +141,13 @@ func (ar *AlertingRule) Exec(ctx context.Context, q datasource.Querier, series b
 	updated := make(map[uint64]struct{})
 	// update list of active alerts
 	for _, m := range qMetrics {
-		for k, v := range ar.Labels {
-			// apply extra labels
+		// extra labels could contain templates, so we expand them first
+		labels, err := expandLabels(m, ar.Labels)
```
Comment (f41gh7) [inline app/vmalert/alerting.go:145]: err check skipped as linter mentioned.

File: app/vmalert/alerting.go [hunk 2]
```diff
@@ -200,6 +202,19 @@ func (ar *AlertingRule) Exec(ctx context.Context, q datasource.Querier, series b
...
+func expandLabels(m datasource.Metric, labels map[string]string) (map[string]string, error) {
+	metricLabels := make(map[string]string)
+	for _, l := range m.Labels {
+		metricLabels[l.Name] = l.Value
+	}
+	tpl := notifier.AlertTplData{
+		Labels: metricLabels,
+		Value:  m.Value,
+	}
+	// we don't expect `query` templates to work here
```
Comment (f41gh7) [inline app/vmalert/alerting.go:214]: in this case, query func for labels wouldn't work at all. Is it worth to add testcase for it and update docs?

File: app/vmalert/alerting.go [hunk 3]
```diff
@@ -200,6 +202,19 @@ func (ar *AlertingRule) Exec(ctx context.Context, q datasource.Querier, series b
...
 }
 
+func expandLabels(m datasource.Metric, labels map[string]string) (map[string]string, error) {
+	metricLabels := make(map[string]string)
+	for _, l := range m.Labels {
+		metricLabels[l.Name] = l.Value
+	}
+	tpl := notifier.AlertTplData{
+		Labels: metricLabels,
+		Value:  m.Value,
```
Comment (f41gh7) [inline app/vmalert/alerting.go:212]: and expr is missing. Will It break labels templating with grafana links?

**Review by f41gh7 — COMMENTED** [review]
Looks good, have few questions.

File: app/vmalert/alerting.go (hunk 2 above)
Comment (hagen1778) [reply app/vmalert/alerting.go]: Yeah, I was thinking maybe it is a not good idea to allow `query` in labels. However, it is up to user.

**Review by f41gh7 — APPROVED** [review]
LGTM

**Review by valyala — APPROVED** [review]

**Comment by valyala** [comment]
Thanks!

**Pull request merged**